### PR TITLE
Add heapster, grafana, influxdb

### PR DESCRIPTION
**Pull request by Cynerva**
WIP, do not merge.

This needs a lot of testing. I'll add more details to the PR on Monday.

### CODE REVIEW

**Comment by Cynerva** [comment]
@chuckbutler This PR adds heapster, grafana, and influxdb to the kubernetes deployment when dashboard=true.

There are two things I want to highlight here.

#### 1. We grabbed definitions from the heapster repo instead of the kubernetes one.

We did this because the [heapster definitions](https://github.com/kubernetes/heapster/tree/master/deploy/kube-config/influxdb) looked newer and easier to work with. The [kubernetes ones](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/cluster-monitoring/influxdb) require rendering some kind of template and there's no supporting build system for it that I can see.

@mbruzek We will need to do further research on making addon updates less of a manual process. Right now it looks like we're gonna have to do some upstream work to make it doable. Sorry we didn't make any headway on that here.

#### 2. We changed the way kubelet is initially brought up

We removed the [step where kubelet is brought up with no DNS config](https://github.com/juju-solutions/kubernetes/pull/65/files#diff-bdfd6f6b79f7f3171e9f0167164a7c98L138).

In order to make this work, we had to [send DNS info before we deploy kube-dns](https://github.com/juju-solutions/kubernetes/pull/65/files#diff-e6f5b987d0e976d7b76acd341abc74c0L157).

The reason we did this was to allow an early deploy of Heapster and friends. Without this change, they came up with missing DNS config and were unable to communicate with eachother.

**Comment by Cynerva** [comment]
Pushed a change to use the heapster templates from the kubernetes repo instead. Looks like that's doable enough.

**Comment by Cynerva** [comment]
@chuckbutler @mbruzek I think I've done about all I can on my own here. This is ready for review and further updates based on feedback.

One problem is that the limits in [heapster-controller.yaml](https://github.com/juju-solutions/kubernetes/blob/ec507a99e1852717a14308c211948d55652c2727/cluster/juju/layers/kubernetes-master/templates/heapster-controller.yaml) are calculated one time during initial deployment, and never updated as new nodes are added or removed. Spent a good chunk of the day trying to get that working with a re-rendered template and `kubectl apply`, but that didn't seem to actually update the limits, so I had to give up on that.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 1]
```diff
@@ -154,8 +154,11 @@ def start_master(etcd, tls):
     set_state('kubernetes-master.components.started')
 
 
-@when('kube-dns.available', 'cluster-dns.connected', 'sdn-plugin.available')
+@when('cluster-dns.connected', 'sdn-plugin.available')
 def send_cluster_dns_detail(cluster_dns, sdn_plugin):
+    ''' Send cluster DNS info '''
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Did you just __add__ a docstring? You are possibly my new best friend!

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 2]
```diff
@@ -227,22 +230,26 @@ def reset_states():
...
-    manifest = '/etc/kubernetes/addons/dashboard.yaml'
+    templates = [
+        'kubernetes-dashboard.yaml',
+        'influxdb-grafana-controller.yaml',
+        'influxdb-service.yaml',
+        'grafana-service.yaml',
+        'heapster-controller.yaml',
+        'heapster-service.yaml'
+    ]
     if hookenv.config('dashboard'):
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Perhaps this configuration name does not make sense any longer. Maybe we need to deprecate and change it to enable-addons or something like that.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 3]
```diff
@@ -265,15 +272,14 @@ def start_kube_dns(sdn_plugin):
         hookenv.log('kube-apiserver not ready, not requesting dns deployment')
         return
 
+    message = 'Rendering the Kubernetes DNS files.'
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: +1 to additional status message here, and log. Good call.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 4]
```diff
@@ -385,41 +391,27 @@ def ceph_storage(ceph_admin):
...
-    return_code = call(split(get))
-    if return_code != 0:
-        # Create the kubedns service from the rendered file.
-        create = 'kubectl create -f {0}/kubedns-svc.yaml'.format(addon_dir)
-        check_call(split(create))
+def create_addon(template, context):
+    '''Create an addon from a template'''
+    target = '/etc/kubernetes/addons/' + template
+    render(template, target, context)
+    cmd = ['kubectl', 'create', '-f', target]
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Creating the same object twice throws an error from kubectl, we may want to check if it is already there with a `kubectl get -f target` if we ever see a case where an attempt is made to the same addon twice.

File: cluster/juju/layers/kubernetes-master/templates/heapster-controller.yaml [hunk 5]
```diff
@@ -0,0 +1,138 @@
+{% set base_metrics_memory = "140Mi" -%}
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-master/templates/heapster-controller.yaml]: This looks like a lot of variables. Did you have to deviate from the original template much here?

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 3 above)
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: This message was originally in the `launch_dns` function which I removed in refactoring. Just preserving the message, it's not a new one!

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 4 above)
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Good point. I think I'd like to use [kubectl apply](http://kubernetes.io/docs/user-guide/kubectl/kubectl_apply/) here instead:

> Apply a configuration to a resource by filename or stdin. This resource will be created if it doesn’t exist yet.

File: cluster/juju/layers/kubernetes-master/templates/heapster-controller.yaml (hunk 5 above)
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/templates/heapster-controller.yaml]: This is an exact copy of the upstream template. I didn't tweak any of the variables.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 2 above)
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: I changed it to `enable-dashboard-addons` for now

**Comment by Cynerva** [comment]
Haven't seen any problems during manual testing with locally built charms.

* Deployed with `bundle-kubernetes-core`, made sure metrics appear in dashboard
* Deployed with `bundle-canonical-kubernetes`, made sure metrics appear in dashboard
* Enabled and disabled the `enable-dashboard-addons` config to ensure it adds and removes dashboard addons properly
* Ran bundletester on `bundle-canonical-kubernetes` with the locally built charms

**Comment by Cynerva** [comment]
Added an incremental update to the kubernetes-master README, but it looks like that README could use some more love. I'd rather address that in a separate PR.

**Comment by mbruzek** [comment]
The README.md file needs more than the definition of the configuration variable. We need to have a section about what the dashboard and heapster are and how to get to them if enabled. That can be in a different PR because as you point out the README needs several updates to catch up with current code.

Great work here I will merge this as is.

**Pull request merged**